### ISSUE TITLE
IsolateClient initialisation is really slow on some devices

### ISSUE THREAD

**Issue by biklas7** [gql-dart/ferry#618]
For some reason with some of our users, the Ferry IsolateClient takes about 20 seconds to initialize.

I actually never was able to replicate this issue on any of my devices, it's actually pretty fast, but using some test builds that logged the client state we were able to determine it was the client itself that takes that long to initialize.

This is a really weird bug and I'm not even sure if it's related to Ferry itself or maybe potentially a bug with Flutter and isolates?
I don't know, but still ,I wanted to create this issue to know if there's anyone here that experienced the same issue or has any insight on how to fix it 🙏

**Comment by knaeckeKami** [comment]
Are you using a persisted cache?

If you have a lot of objects in your cache you might want to evict() some old queries/mutation and call .gc()

**Comment by biklas7** [comment]
> Are you using a persisted cache?
Yes! The user helping us test this has lots of objects cached. We never clear or do anything with the cache, just when a user logs out. We asked him to do just that, and after a new login, the client was able to initialize again faster, so this seems to be the issue.

My question now is, what do you recommend to do? Are there any best practices to avoid this? Like clearing the cache from time to time or maybe something else? 🤔

**Comment by biklas7** [comment]
I see 🤔 thanks for the comprehensive answer @knaeckeKami 🙏 

We do have some queries with arguments that constantly change, either some that take a date or search by some string, but there are not many so probably not worth the effort of manually evicting previous ones, because we do not keep track of previously used arguments anyway. Or is there a way to say to evict everything but a specific query?

It does make sense to evict the Mutations, we do cache them because let's say we have an edit operation that returns the updated item, we want to cache that. But wouldn't calling `.gc()` also clear the mutations?

And talking about garbage collection, it makes perfect sense what you say, but why doesn't the client do this automatically behind the scenes?

**Comment by knaeckeKami** [comment]
> It does make sense to evict the Mutations, we do cache them because let's say we have an edit operation that returns the updated item, we want to cache that. But wouldn't calling .gc() also clear the mutations?

Just calling gc() is not enough, because the mutation field itself will be cached

e.g.

```graphql
muation EditUser {
   editUser(firstName: "Jack") {
      id
      firstName
    }
}
```

would be cached also as

```json
{
  "Query": {...},
  "Mutation": {
     "editUser(firstName:\"Jack\")": {
           { "__ref": "User:42" }
    }

   },
   "User:42": {
            "__typename": "User",
    "id": "42",
    "firstName": "Jack",
     ...
 }
```

since the `editUser` mutation is reachable from the Root Mutation field, nothing would be garbage collected here. 


> And talking about garbage collection, it makes perfect sense what you say, but why doesn't the client do this automatically behind the scenes?

I'm open to having ferry adopt a default strategy on when to garbage collect, but there is IMO no totally obvious answer on when to run it. 

>  Or is there a way to say to evict everything but a specific query?

Nothing built-in at the moment.
But you can call 
```dart
final queryKeys = client.cache.store.get("Query")?.keys; // or whatever your root query type name is, e.g "RootQuery"
```

and then filter the query fields that you want to remove, and then call

```dart
 for(final key in queryKeysToRemove) {
     client.cache.evict("Query", fieldName: key);
 }
```

**Comment by biklas7** [comment]
Now I get it! Thank you 👍 

> I'm open to having ferry adopt a default strategy on when to garbage collect, but there is IMO no totally obvious answer on when to run it.
> 
I think it would be cool to have something like that, even if a little bit opinionated. Even if it wasn't on by default, it would be nice to have a default strategy for eviction and garbage collection that Ferry users could easily and optionally activate or, if needed, even extend and modify to more specific needs.

I feel like I have everything I need to improve my implementation though. I'll be working on this in the upcoming weeks and if I have more questions I'll add them here, if not I'll close this issue.

Thanks again for explaining everything so well 🙏

**Comment by knaeckeKami** [comment]
Ferry Cache and IsolateClient now support `evictOperation()` methods, which should make it easier to remove old operations.

**Comment by biklas7** [comment]
I'm testing out some of your suggestions and so far it's looking promising, let's see how it works out with our hardcore users that have a big cache.

If anyone finds this useful, we're garbage-collecting the cache, evicting all mutations (shorter intervals), and clearing the entire cache (longer intervals) from time to time.

Thanks again @knaeckeKami for all the help and detailed insights on how the ferry cache actually works 👍
